### DISCUSSION TITLE
Standardize media metadata, add to PlayMedia

### DISCUSSION THREAD

**Discussion by chishm** [home-assistant/architecture#739] [Entity Models]
## Context
There are multiple places in Hass where media (audio, video) is interacted with. Some of these places have rich metadata associated with the media, others do not:
* Currently playing media has `title`, `album_artist`, `album_name`, `duration`, etc. available via `media_player` entity attributes
* The `media_player.play_media` service takes [metadata](https://www.home-assistant.io/integrations/media_player/#metadata) via the `extra` parameter, deferring to metadata classes defined for [Google Cast](https://developers.google.com/cast/docs/reference/messages#MediaData).
* The `PlayMedia` returned when a media source resolves an item does not support any metadata.

## Problems
The metadata isn't consistent. It's difficult to know what can be provided in a `media_player.play_media` call, and it's buried under an `extra` parameter.

Media sources cannot provide any metadata to media players. Some media players require the metadata when an item is played, to reflect back to Hass or the device's own UI.

Extending metadata fields in one location (e.g. media player attributes) won't be automatically extended in the others (e.g. `play_media` calls), leading to repeated inconsistencies.

## Proposal
I'd like to make metadata consistent in the following way:

1. Standardize what media metadata can be provided in the 3 places outlined above.
    * The metadata should be based on what `media_player` already defines, with any useful additions from the Google Cast docs.
2. Turn this standard into user and developer documentation.
3. Create a `MediaMetadata` dataclass where all fields (except maybe title) are optional.
3. Add `metadata` as a top-level (not under `extra`), optional, service call attribute to the `media_player.play_media` service.
4. Deprecate the old `extra.metadata` attribute and automatically move it to the `metadata` attribute for the deprecation period. 
5. Add to `PlayMedia` a `metadata: MediaMetadata | None = None` field for `media_source.async_resolve_media` to fill-in.
6. Maybe move the `media_player` metadata attributes into an explicit `media_metadata` attribute (I'm not decided on the utility of this change yet)

**Comment by balloob** [comment]
There are two integrations that deal with media: `media_player` and `media_source`. The media player besides it's entity also has a `BrowseMedia` class that represents browsable media (parent or child).

The `media_source` integration has `PlayMedia` model: it contains URL and mime type of an item so it can be played. It is returned when a media player is asked to play a media source and wants to turn the media source identifier into something useful. I have [a PR open](https://github.com/home-assistant/core/pull/67009) that will align `PlayMedia` with `BrowseMedia`. That way a media player can show some more information.

The browsable media model has taken a simplified approach to media and reduced everything to just `title` + `thumbnail`. The reason for this is that there are just too many media types and variations, and it makes everything complicated.

Now my above linked PR is not going to solve metadata for all playable items, just for the ones that come from a media source, so I see a need for specifying metadata. However, I wonder if we should just limit this to `title` and `thumbnail`.

**Comment by stboch** [comment]
@chishm Thanks for creating this I was actually bringing up this discussion on the discord channel. 
@balloob While I understand the desire to keep the number limited, it seems that the media_player at least from a status prospective already has much of this data. I don't think its to unreasonable that all the supported types of attributes would also be a available all be it not required argument.  [Media_Player Entity Attributes](https://github.com/home-assistant/core/blob/bdcdf5222549710aeb70266b35d27cd7dbfef6ce/homeassistant/components/media_player/__init__.py#L404)

**Comment by chishm** [comment]
@balloob: Your PR looks like a good starting point, but I think there's still useful metadata missing that could otherwise be supplied from media source to player. As @stboch mentioned, the `media_player` entity's `media_*` attributes would be the most useful set to standardise on (I'll call that the media player (MP) metadata set). We could limit the official metadata to that, and leave anything else as `media_player` integration-specific extensions (as it sort of is now with `cast` and `extra.metadata`), although extensions wouldn't be encouraged.

The reason I have for wanting the MP metadata set is that some media players require the metadata to be supplied when playing a media URL, and won't extract the metadata on their own. Without the metadata, the Hass front-end media player card has missing information. I've skirted the problem in the specific `dlna_dms` (media source) to `dlna_dmr` (media player) case by extending `PlayMedia` with `didl_metadata`, but this doesn't solve it for other cases e.g.: `dlna_dms` to `cast`, or local media to `dlna_dmr`.

Incidentally, if the MP metadata set were implemented for `PlayMedia`, I could drop the `didl_metadata` extension and just have `dlna_dmr.play_media` recreate the DIDL-Lite metadata from the `PlayMedia` metadata. This would reduce complexity in this case and just work for all other media sources that supply metadata too.

Additionally, standardizing a `media_player.play_media` `metadata` parameter would reduce complexity for integrations implementing it, as the code author would not need to consult the Google docs and `cast` integration to figure out what the useful keyword arguments are; they'd be documented in the Hass developer docs and in a `MediaMetadata` class. And users would have a clearer understanding of what metadata arguments they can provide.

I think the `BrowseMedia` class can stay as-is, given it is only used for selecting media, and is not used directly for sending URLs and metadata to media players.

**Reply by balloob** [reply]
Please use threading when giving answers instead of adding new answers to the bottom.

If we are to add a new MediaPlayerMetadata class it should:

- Follow the same categories as cast, so we have 5 different metadata classes
- Be used internally too. So MediaPlayer gets a helper to return an object that holds its current attributes 
- Be accepted as a `metadata` dict when playing media
- Can be attached to `PlayItem` model

So what would those items look like?

**Reply by chishm** [reply]
Sorry about not threading, I wasn't sure of the etiquette for this forum.

Here's a first pass at the metadata class (at bottom of this post).

I've deliberately kept it flat (instead of having 5 derived classes) to keep it simple to use. Implementations of `MediaPlayer.play_media` can just use an attribute without having to check the class type first. Service calls can be made with a `metadata` argument which is parsed directly into a `MediaMetadata` object.

The attributes are mostly the same as `MediaPlayer.media_*` attributes, with some extras taken from the [cast metadata classes](https://developers.google.com/cast/docs/reference/messages#GenericMediaMetadata). I haven't added a `playlist_title`, as the `title` attribute would apply if the media item is a playlist.

I wasn't entirely sure whether to have `mime_type` included, or just rely on the holder of the metadata object to already know (e.g. via `MediaPlayer.media_content_type` or `PlayMedia.mime_type`). Having it here makes the metadata object self-describing (knows what type the media is and therefore what fields should be available), but it adds redundancy.

```py
@dataclass
class MediaMetadata:
    """Data describing a media item."""

    # Generic metadata (applies to all types)
    title: str = "Home Assistant"
    mime_type: str
    
    release_date: datetime | None = None
    
    # Thumbnail image
    thumbnail_url: str | None = None
    thumbnail_height: int | None = None
    thumbnail_width: int | None = None
    thumbnail_remotely_accessible: bool = False

    # Generic to linear media (e.g. music, movies)
    position: int | None = None
    duration: int | None = None

    # Movie metadata
    # (Nothing extra)
    
    # TV Show metadata
    series_title: str | None = None
    season: str | None = None
    episode: str | None = None
    
    # Music track metadata
    album_name: str | None = None
    album_artist: str | None = None
    artist: str | None = None
    track: int | None = None

    # Photo metadata
    location: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    height: int | None = None
    width: int | None = None
    # (Artist is shared with music track metadata)
```

How's it looking so far?

**Reply by stboch** [reply]
A track or album might be a part of a playing playlist. So theoretically it could still be populated while the title is the Track title playing. 

mime_type will need to be included as it is needed for some players. and probably good to to still include the media_type_class unless there is another place that would be defined while I know it adds redundancy I am not sure that media_source would pass along that information otherwise which might be necessary for a player to decide which path to take for playing a class of media.

**Reply by balloob** [reply]
We should not come up with new fields. The proposal was to use the existing media player entity properties, so that's what we should limit it to. But we should still group it in different groups based on what the existing properties document that [they are limited to](https://github.com/home-assistant/core/blob/dev/homeassistant/components/media_player/__init__.py#L542).

**Reply by chishm** [reply]
Thanks for the feedback. I've stuck with fields (with original names) only defined in media player entities, including playlist. I've also dropped the mime_type as it is specified at the top level in every case.

Here's the revised class:
```py
@dataclass
class MediaMetadata:
    """Data describing a media item."""

    # Generic metadata (applies to all types)
    title: str | None = None
    
    # Thumbnail image
    image_url: str | None = None
    image_remotely_accessible: bool = False

    # Generic metadata for linear media (e.g. music, movies)
    duration: int | None = None
    position: int | None = None
    
    # Music track metadata
    artist: str | None = None
    album_name: str | None = None
    album_artist: str | None = None
    track: int | None = None
    playlist: str | None = None
    
    # TV Show metadata
    series_title: str | None = None
    season: str | None = None
    episode: str | None = None
    channel: str | None = None

    # Movie metadata
    # (Nothing extra)

    # Photo metadata
    # (Nothing extra)
```

Here's the `play_media` signature to accept it. Note metadata is keyword-only for compatibility with existing `MediaPlayer`-derived classes.
```py
def play_media(self, media_type: str, media_id: str, *, metadata: MediaMetadata, **kwargs):
    """Play a piece of media."""
```

Here's the `PlayMedia` class. Again metadata is optional for compatibility with existing media sources.
```py
@dataclass
class PlayMedia:
    """Represents a playable media."""

    url: str
    mime_type: str
    metadata: MediaMetadata | None = None
```

**Reply by balloob** [reply]
Note that `position` is not metadata. It's part of the state of playing the item. We also don't store `position` in Home Assistant but tie it to a timestamp, as to not have to constantly update the state machine.

**Reply by balloob** [reply]
I still think that we should try to break up the metadata in different classes so the integration doesn't have to guess which fields to use. Probably need a `metadata_type`

**Reply by chishm** [reply]
I'd included `position` in the metadata based on it being used as a bookmark partway into a stream (e.g resuming a movie 20 minutes in), but you're right that it's not really metadata and is better done using a seek after playing.

Here's how it could look if broken up into classes based on media type. I'm not sure how to easily parse it from a flat `metadata` argument in a `media_player.play_media` service call.

```py
@dataclass
class MediaMetadata:
    """Data describing a media item."""

    # Generic metadata (applies to all types)
    title: str | None = None
    
    # Thumbnail image
    image_url: str | None = None
    image_remotely_accessible: bool = False

@dataclass
class MusicMediaMetadata(MediaMetadata):
    """Data describing a music track."""
    
    duration: int | None = None
    artist: str | None = None
    album_name: str | None = None
    album_artist: str | None = None
    track: int | None = None
    playlist: str | None = None
   
@dataclass
class TVShowMediaMetadata(MediaMetadata):
    """Data describing a TV Show."""
  
    duration: int | None = None
    series_title: str | None = None
    season: str | None = None
    episode: str | None = None
    channel: str | None = None

@dataclass
class MovieMediaMetadata(MediaMetadata):
    """Data describing a movie."""
    
    duration: int | None = None

@dataclass
class ImageMediaMetadata(MediaMetadata):
    """Data describing a photograph or image."""
    
    artist: str | None = None
```

**Reply by chishm** [reply]
Regarding using a `metadata_type` field, are we trying to supplement or replace the `media_player.media_content_type` field?

Should it follow:
    * [`media_player.const.MEDIA_CLASS_*`](https://github.com/home-assistant/core/blob/0bd25a192d4eb352bd4e0b105a31ca423923b77d/homeassistant/components/media_player/const.py#L37) constants,
    * [`MEDIA_TYPE_*` constants,](https://github.com/home-assistant/core/blob/0bd25a192d4eb352bd4e0b105a31ca423923b77d/homeassistant/components/media_player/const.py#L58)
    * Defined values for `media_content_type` in calls to [media_player.play_media](https://www.home-assistant.io/integrations/media_player/#service-media_playerplay_media),
    * [Google Cast `metadataType`](https://developers.google.com/cast/docs/reference/messages#GenericMediaMetadata) values, or
    * MIME types?

None of the above map easily into any of the others. When playing to a device, the MIME type is often preferred for indicating the type of media. For selecting metadata, we could follow the hierarchy set out above, which works well for `cast` but maybe not everything else, or we could duck-type it and just use any available and relevant fields.

For comparison, [here is how dlna_dms](https://github.com/home-assistant/core/blob/dev/homeassistant/components/dlna_dmr/const.py#L71) maps from HA's `media_content_type` to a suitable DIDL-Lite class, and [how it translates metadata field names](https://github.com/home-assistant/core/blob/0bd25a192d4eb352bd4e0b105a31ca423923b77d/homeassistant/components/dlna_dmr/const.py#L97) from Google Cast's definitions to DIDL-Lite (where field names need translation; others are kept-as is (e.g. `artist` and `album_name`).

**Reply by balloob** [reply]
`media_content_type` is a MIME type if it's pointing at a file. If it's pointing at library items it only knows how to play itself (ie Spotify), it can use the MEDIA_TYPE constants or make up what it wants

**Reply by balloob** [reply]
But yeah it sucks to introduce yet another type. Maybe it should be matched to media class?

**Reply by chishm** [reply]
Matching to media class sounds like the right idea to me. It's a well-defined set that describes most (if not all) cases. It can then be mapped to the set that the integration needs (like the 5 metadata classes for `cast`, or the DIDL-Lite classes for `dlna_dmr`).

 I've expressed the restriction of `media_class` values using a `StrEnum`, but I'm not sure how well mypy handles it. At runtime we can catch value errors and log for hints to the user, rather than letting invalid `media_class` values slip through service calls and into integrations that don't expect them.

All fields in `MediaMetadata` are still optional, including `media_class`. It's provided on a best-effort basis of "here's what information we know" without forcing sources (or users) to make something up just to add metadata.

I've gone with the flat metadata class, since the fields apply in many different combinations to the various `media_class`es.

```py
class MediaClass(StrEnum):
    """All valid media class values."""

    ALBUM = MEDIA_CLASS_ALBUM
    APP = MEDIA_CLASS_APP
    ARTIST = MEDIA_CLASS_ARTIST
    CHANNEL = MEDIA_CLASS_CHANNEL
    COMPOSER = MEDIA_CLASS_COMPOSER
    CONTRIBUTING = MEDIA_CLASS_CONTRIBUTING
    DIRECTORY = MEDIA_CLASS_DIRECTORY
    EPISODE = MEDIA_CLASS_EPISODE
    GAME = MEDIA_CLASS_GAME
    GENRE = MEDIA_CLASS_GENRE
    IMAGE = MEDIA_CLASS_IMAGE
    MOVIE = MEDIA_CLASS_MOVIE
    MUSIC = MEDIA_CLASS_MUSIC
    PLAYLIST = MEDIA_CLASS_PLAYLIST
    PODCAST = MEDIA_CLASS_PODCAST
    SEASON = MEDIA_CLASS_SEASON
    TRACK = MEDIA_CLASS_TRACK
    TV = MEDIA_CLASS_TV
    URL = MEDIA_CLASS_URL
    VIDEO = MEDIA_CLASS_VIDEO

@dataclass
class MediaMetadata:
    """Data describing a media item."""

    # Class of media that this metadata describes
    media_class: MediaClass | None = None
    
    # Generic metadata (applies to all types)
    title: str | None = None
    
    # Thumbnail image
    image_url: str | None = None
    image_remotely_accessible: bool = False

    # Generic metadata for linear media (e.g. music, movies)
    duration: int | None = None
    
    # Music (and other audio) metadata
    artist: str | None = None
    album_name: str | None = None
    album_artist: str | None = None
    track: int | None = None
    playlist: str | None = None
    
    # TV Show metadata
    series_title: str | None = None
    season: str | None = None
    episode: str | None = None
    channel: str | None = None

    # Movie metadata
    # (Nothing extra)

    # Photo (and other images) metadata
    # (Nothing extra)
```
